XMonad/Layout/DecorationEx/Engine.hs

```diff
@@ -0,0 +1,511 @@
+                                                        -- which are to be used for this particular window in it's current state.
+  , ddOrigWindow :: !Window                             -- ^ Original window to be decorated
+  , ddWindowTitle :: !String                            -- ^ Original window title (not shrinked yet)
+  , ddDecoRect :: !Rectangle                            -- ^ Decoration rectangle
+  , ddWidgets :: !(WidgetLayout widget)         -- ^ Widgets to be placed on decoration
+  , ddWidgetPlaces :: !(WidgetLayout WidgetPlace)       -- ^ Places where widgets must be shown
+  }
+
+-- | State of decoration engine
+data DecorationLayoutState engine = DecorationLayoutState {
+    dsStyleState :: !(DecorationEngineState engine) -- ^ Engine-specific state
+  , dsDecorations :: ![WindowDecoration]            -- ^ Mapping between decoration windows and original windows
+  }
+
+-- | Decoration engines type class.
+-- Decoration engine is responsible for drawing something inside decoration rectangle.
+-- It is also responsible for handling X11 events (such as clicks) which happen
+-- within decoration rectangle.
+-- Decoration rectangles are defined by DecorationGeometry implementation.
+class (Read (engine widget a), Show (engine widget a),
+       Eq a,
+       DecorationWidget widget,
+       HasWidgets (Theme engine) widget,
+       ClickHandler (Theme engine) widget,
+       ThemeAttributes (Theme engine widget))
+    => DecorationEngine engine widget a where
+
+    -- | Type of themes used by decoration engine.
+    -- This type must be parametrized over widget type,
+    -- because theme will contain a list of widgets.
+    type Theme engine :: Type -> Type           
+                                          
+    -- | Type of data used by engine as a context during painting;
+    -- for plain X11-based implementation this is Display, Pixmap
+    -- and GC.
+    type DecorationPaintingContext engine 
+ 
+    -- | Type of state used by the decoration engine.
+    -- This can contain some resources that should be initialized
+    -- and released at time, such as X11 fonts.
+    type DecorationEngineState engine     
+
+    -- | Give a name to decoration engine.
+    describeEngine :: engine widget a -> String
+
+    -- | Initialize state of the engine.
+    initializeState :: engine widget a       -- ^ Decoration engine instance
+                    -> geom a                -- ^ Decoration geometry instance
+                    -> Theme engine widget   -- ^ Theme to be used
+                    -> X (DecorationEngineState engine)
+
+    -- | Release resources held in engine state.
+    releaseStateResources :: engine widget a              -- ^ Decoration engine instance
+                          -> DecorationEngineState engine -- ^ Engine state
+                          -> X ()
+
+    -- | Calculate place which will be occupied by one widget.
+    -- This method is expected to always return a rectangle which has X = 0,
```

Comment:
Agreed. But Point is not enough here. Basically, I need almost full Rectangle, i.e. I need Y coordinate of left-top corner, width and height. X coordinate is not needed, because I will then move this rectangle to left or to right in order to align widgets properly. Maybe declare special data type for this case?